File: Linear/Dynamic/Linked List/Circular Linked List/02_Deleting/End/Deleting.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

def deleteAtEnd(last):
    head = last.next

    if last is None or last is head:
        return head
    
    secondLast = head
    while secondLast.next is not last:
        secondLast = secondLast.next
    
    secondLast.next = head

    last = secondLast

    return last

def printList(last):
    if last is None:
        return
    
    head = last.next
    temp = head

    while True:
        print(temp.data, end="")
        temp = temp.next
        if temp is not head:
            print(" -> ", end="")
        else:
            break
    
    print()

# Create circular linked list: 2 -> 3 -> 4
first = Node(2)
first.next = Node(3)
first.next.next = Node(4)

last = first.next.next
last.next = first

print("Linked List before deletion\n")
printList(last)

last = deleteAtEnd(last)

print("Linked List after deletion\n")
printList(last)

File: Linear/Dynamic/Linked List/Circular Linked List/02_Deleting/End/Deleting.java
class Node {
    int data;
    Node next;

    Node(int data) {
        this.data = data;
        this.next = null;
    }
}

class DeletingEnd {
    public static Node deleteAtEnd(Node last) {
        Node head = last.next;

        if (last == null || last == head) {
            return null;
        }

        Node secondLast = head;
        while (secondLast.next != last) {
            secondLast = secondLast.next;
        }

        secondLast.next = head;

        last = secondLast;

        return last;
    }

    public static void printList(Node last) {
        if (last == null)
        return;

        Node head = last.next;
        Node temp = head;

        do {
            System.out.print(temp.data);
            temp = temp.next;
            if (temp != head)
            System.out.print(" -> ");
        } while (temp != head);

        System.out.println();
    }

    public static void main(String[] args) {
        // Create circular linked list: 2 -> 3 -> 4
        Node first = new Node(2);
        first.next = new Node(3);
        first.next.next = new Node(4);

        Node last = first.next.next;
        last.next = first;

        System.out.println("Linked List before deletion");
        printList(last);

        last = deleteAtEnd(last);

        System.out.println("Linked List after deletion");
        printList(last);
    }
}